kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiBatchApplicationStream.java

```diff
@@ -88,27 +91,83 @@ public void preShutdown() {
+                }
+
+                logger.trace("Sending records to Storage API writer for batch load...");
+                ApiFuture<AppendRowsResponse> response = applicationStream.writer().append(jsonRecords);
+                AppendRowsResponse writeResult = response.get();
+                logger.trace("Received response from Storage API writer batch...");
+
+                if (writeResult.hasAppendResult()) {
+                    logger.trace("Append call completed successfully on stream {}", streamName);
+                    updateSuccessAndTryCommit(applicationStream, tableName, streamName);
+                    return;
+                } else if (writeResult.hasUpdatedSchema()) {
+                    logger.warn("Sent records schema does not match with table schema, will attempt to update schema");
+                    retryHandler.attemptTableOperation(schemaManager::updateSchema);
+                } else if (writeResult.hasError()) {
+                    Status errorStatus = writeResult.getError();
+                    String errorMessage = String.format("Failed to write rows on table %s due to %s", tableName, errorStatus.getMessage());
+                    retryHandler.setMostRecentException(new BigQueryStorageWriteApiConnectException(errorMessage));
+                    if (BigQueryStorageWriteApiErrorResponses.isMalformedRequest(errorMessage)) {
+                        rows = mayBeHandleDlqRoutingAndFilterRecords(rows, convertToMap(writeResult.getRowErrorsList()), tableName.getTable());
+                        if (rows.isEmpty()) {
+                            updateSuccessAndTryCommit(applicationStream, tableName, streamName);
+                            return;
+                        }
+                    } else if (!BigQueryStorageWriteApiErrorResponses.isRetriableError(errorStatus.getMessage())) {
+                        // Fail on non-retriable error
+                        logger.error(errorMessage);
+                        throw retryHandler.getMostRecentException();
+                    }
+                    logger.warn(errorMessage + " Retry attempt " + retryHandler.getAttempt());
+                }
+            } catch (BigQueryStorageWriteApiConnectException exception) {
+                throw exception;
+            } catch (Exception e) {
+                String message = e.getMessage();
+                String errorMessage = String.format("Failed to write rows on table %s due to %s", tableName, message);
+                retryHandler.setMostRecentException(new BigQueryStorageWriteApiConnectException(errorMessage, e));
+
+                if (canAttemptSchemaUpdate()
+                        && ((BigQueryStorageWriteApiErrorResponses.isMalformedRequest(message)
+                        && BigQueryStorageWriteApiErrorResponses.hasInvalidSchema(getRowErrorMapping(e).values()))
+                        || BigQueryStorageWriteApiErrorResponses.hasInvalidSchema(Collections.singletonList(errorMessage)))) {
```

Comment:
the conditions in the if clause can be abstracted into a separate method for better readability

Reply:
same with else clauses subsequently

Reply:
done